front-end/src/visualization/Visualization.jsx

```diff
@@ -50,47 +57,115 @@ function Visualization() {
-        const clusterColor = getRandomColor();
-        const songPositions = calculateCircularPositions(centerX, centerY, radius, songCount);
-        const nodes = [{ id: clusterId, label: clusterLabel, x: centerX, y: centerY, isCluster: true, color: clusterColor }];
-        const edges = [];
+    function generateClusterData(clusterData, clusterPositions) {
+        let globalId = 1;
+        let allNodes = [];
+        let allEdges = [];
 
-        songPositions.forEach((pos, index) => {
-            const songId = clusterId * 100 + index + 1;
-            nodes.push({ id: songId, label: `Song ${index + 1}`, x: pos.x, y: pos.y, color: clusterColor });
-            edges.push({ from: clusterId, to: songId, color: 'white' });
+        clusterData.centroids.forEach((centroid, index) => {
+            const formattedCentroid = clusterData.centroids[index]?.toFixed(2) || '0.00';
+            const clusterLabel = `Valence: ${formattedCentroid}`;
+            const { x, y } = clusterPositions[index];
+            const songsInCluster = clusterData.assignments.filter(song => song.cluster === index);
+            const songPositions = calculateCircularPositions(x, y, 100, songsInCluster.length);
+            const clusterColor = getRandomColor();
+
+            const nodes = [{ id: globalId++, label: clusterLabel, x, y, isCluster: true, color: clusterColor }];
+            const edges = [];
+
+            songsInCluster.forEach((song, idx) => {
+                nodes.push({
+                    id: song.id,
+                    label: song.name,
+                    x: songPositions[idx].x,
+                    y: songPositions[idx].y,
+                    color: clusterColor,
+                    imageUrl: song.album.images[0].url
+                });
+                edges.push({ from: globalId - 1, to: song.id, color: 'white' });
+            });
+
+            allNodes = allNodes.concat(nodes);
+            allEdges = allEdges.concat(edges);
         });
 
-        return { nodes, edges };
+        return { nodes: allNodes, edges: allEdges };
     }
 
-    function addInterClusterEdges(nodes, edges) {
-        const song1 = nodes.find(node => node.label === 'Song 1' && node.id > 100 && node.id < 200);
-        const song3 = nodes.find(node => node.label === 'Song 3' && node.id > 200 && node.id < 300);
-        if (song1 && song3) {
-            edges.push({
-                from: song1.id,
-                to: song3.id,
-                color: 'red'
-            });
+    useEffect(() => {
+        const getClusterData = async () => {
+            try {
+                const response = await fetch(`${import.meta.env.VITE_BACKEND_ADDRESS}/clusters/${stitchId}`);
+                const data = await response.json();
+                setClusterData(data);
+            } catch (error) {
+                console.error('Error getting clusters:', error);
+            }
```

Comment:
I feel I've raised a considerable amount of times how it is better to actually handle the error instead of doing a `console.error`, which will just log information to the console. Would you mind letting me know why are you not handling the errors and following the `console.error` pattern?

Let's take this case as an example. If the `getClusterData` function fails to fetch the information, do we notify the user in any shape or form? From what I see we're not doing anything, while we do have the `catch` part to know something went wrong, and raise something as simple as a string with "Failed to generate clustering information".

Reply:
> I feel I've raised a considerable amount of times how it is better to actually handle the error instead of doing a `console.error`, which will just log information to the console. Would you mind letting me know why are you not handling the errors and following the `console.error` pattern?
> 
> Let's take this case as an example. If the `getClusterData` function fails to fetch the information, do we notify the user in any shape or form? From what I see we're not doing anything, while we do have the `catch` part to know something went wrong, and raise something as simple as a string with "Failed to generate clustering information".

Yes, we haven't discussed what to do instead so I can adjust all of them. Let's talk about in our 1:1.